presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java

```diff
@@ -96,6 +103,25 @@ private static OptionalLong reduce(OptionalLong first, OptionalLong second, Redu
         return OptionalLong.empty();
     }
 
+    public static <T> List<T> pickRandomSample(List<T> elements, int sampleSize)
```

Comment:
The problem with this is query execution performance may vary (become non deterministic) and also plan itself will be non deterministic. It would be nice to always choose the same sample partitions for a table. What do you think?

Reply:
I was thinking about having a deterministic pseudo-random picking algorithm. The idea is to compute a hash from a partition name, than sort the partitions by hash and get a sub list. But that might be expensive though. Let me do the benchmarks first.

Reply:
What about taking `elements.size()` as the only seed?

Also, if I had 100 sample size of partitions, I would always take a look at first & last partitions, to capture min/max for columns that are correlated with partition key.

Reply:
> > What about taking elements.size() as the only seed?
> I didn't get this

To use `new Random(partitions.size())` or something like that.
Or, in fact, just `new Random(42)` -- I am not convinced this query-to-query randomness buys us anything.

> It is hard to reason what is the first and last partition. Metastore is not guaranteed to return the partition in any particular order. Also the lexicographical order of the partition names does not necessary matches the horological / natural order.

Good observation. Especially if there are multiple partitioning keys.
Still, I consider this just a heuristic, taking a chance, shouldn't hurt. I would do this only if sample size is above some (arbitrarily chosen) threshold, eg 50

Reply:
Actually yes, that make sense. I'm going to introduce min-max sampling. I'm not going to introduce magic threshold though. If the number of samples is very small - the result won't be good anyway, no matter how do you select the sample.

Reply:
Indeed. Just curious about the edge case -- what will you pick when sample size is 1?

Reply:
I will pick the minimum partition in the lexicographical order. But setting the sample size to `1` is an unlikely scenario.